### ISSUE TITLE
[Request] Add window reactions and alerts to support for wayland

### ISSUE THREAD

**Issue by olivertzeng** [Monika-After-Story/MonikaModDev#10336]
are there any limitations to port this feature to wayland? Also for alerts, doesn't `notify-send` work just fine on wayland?

I'm talking about this part of the code:
```python
if session_type == "wayland":
 	store.mas_windowreacts.can_show_notifs = False
 	store.mas_windowreacts.can_do_windowreacts = False
	store.mas_utils.mas_log.warning("Wayland is not yet supported, disabling notifications.")
```

**Comment by dreamscached** [comment]
As of the latest MAS update Wayland did not provide an interface for requesting window titles, hence barring us from having a way to react to a window title.

If that changed or I have overlooked something, it'd be great if you shared an approach.

notify-send works by using libnotify interfaces desktop environments use, it can work on Wayland.

**Comment by olivertzeng** [comment]
> As of the latest MAS update Wayland did not provide an interface for requesting window titles, hence barring us from having a way to react to a window title.
> 
> If that changed or I have overlooked something, it'd be great if you shared an approach.
> 
> notify-send works by using libnotify interfaces desktop environments use, it can work on Wayland. Why it is disabled I don't know, but I speculate it's because it is primarily (or originally was) used for window reactions only.

Hey! thanks for the quick reply!
yes, there is in fact a tool that works! It's called kdotool

https://old.reddit.com/r/kde/comments/1c447nc/anything_like_xprop_in_wayland_to_get_window/kzlaur9/

![圖片](https://github.com/user-attachments/assets/d1b2c2de-4661-41ce-8ff7-39ad27acd3bb)
It's surprising about how it worked though, fascinating

**Comment by olivertzeng** [comment]
but it doesn't show the webpage though, sad

**Comment by multimokia** [comment]
The original implementation of windowreacts and mouse following used xdotool instead of the current xlib system. 

The issue with this, is that for something like mouse position checking, it's extremely expensive as we need to poll a terminal process, which adds notable lag/latency during idle.

I don't see a reason it would be different with kdotool either. I'd also like to investigate if that works on more than just KDE. (Going by the name, at least.)

If we did use this for Wayland, then it would make sense to keep mouse following and such inactive for performance

**Comment by olivertzeng** [comment]
> The original implementation of windowreacts and mouse following used xdotool instead of the current xlib system.
> 
> The issue with this, is that for something like mouse position checking, it's extremely expensive as we need to poll a terminal process, which adds notable lag/latency during idle.
> 
> I don't see a reason it would be different with kdotool either. I'd also like to investigate if that works on more than just KDE. (Going by the name, at least.)
> 
> If we did use this for Wayland, then it would make sense to keep mouse following and such inactive for performance

you can disable that flag though

the command i ran was
`watch -n.1 kdotool getmouselocation getwindowpid getwindowclassname getwindowname`
which I guess to only display the window name you could just do
`watch -n.1 kdotool getwindowclassname` instead

the problem is that it can only fetch the **title** of the app instead of the url which is kinda annoying, like i.e. searching Reddit on Google would result in a conflict

**Comment by olivertzeng** [comment]
https://stackoverflow.com/questions/18436245/how-to-fetch-url-of-current-tab-in-my-chrome-extension-using-javascript

or maybe try to have an additional addon that the user could use to use window reactions

**Comment by multimokia** [comment]
It's more that we spawned a process during polling. I suppose we could maintain the process and poll its output.

We are not looking to create external solutions. So no chrome/Firefox extensions will be made

**Comment by olivertzeng** [comment]
So a Python module is much more practical than a command line tool it would seem.

**Comment by olivertzeng** [comment]
https://unix.stackexchange.com/questions/629440/how-to-get-the-url-from-current-tab-of-firefox-without-need-to-install-addition

> I wouldn't use xdotool for that - it is too fragile and error-prone. Use a proper browser automation tool instead such as [Marionette](https://firefox-source-docs.mozilla.org/testing/marionette/Intro.html). Install marionatte_driver python module:
> 
> `pip3 install --user marionette_driver`
> 
> Start a new instance firefox with --marionette option:
> 
> `firefox --marionette`
> 
> Use the following get_url.py:
> ```python
> #!/usr/bin/env python3
> 
> import marionette_driver
> 
> m = marionette_driver.marionette.Marionette()
> 
> m.start_session()
> print(m.get_url())
> m.delete_session()
> ```
> 
> Example:
> ```sh
> $ ./get_url.py
> https://unix.stackexchange.com/questions/629440/how-to-get-the-url-from-current-tab-of-firefox-by-help-of-wmctrl
> ```
> 
> You can save output of ./get_url.py to a variable using command substitution:
> 
> ```sh
> $ url="$(./get_url.py )"
> $ echo $url
> https://unix.stackexchange.com/questions/629440/how-to-get-the-url-from-current-tab-of-firefox-by-help-of-wmctrl/629447?noredirect=1#comment1177925_629447
> ```

Kinda cool solution, but apprently only works on gecko browsers

**Comment by olivertzeng** [comment]
Oh wait

> Does it need to do any additional like open the FF browser on a special way before ? I dont get any output by this on Linux Debian/Ubuntu. – 
> [Alfred.37](https://unix.stackexchange.com/users/412111/alfred-37)
> Commented Jan 26, 2021 at 23:52
> @Alfred.37: yes, you need, and I described it. – 
> [Arkadiusz Drabczyk](https://unix.stackexchange.com/users/72304/arkadiusz-drabczyk)
> Commented Jan 27, 2021 at 10:15

that's actually a kind of problem but im sure by telling what the users should do to enable this it should be fine ig

**Comment by olivertzeng** [comment]
![圖片](https://github.com/user-attachments/assets/3d7f9c3a-43c0-4ebe-948f-83d4845b7b9e)
it works

**Comment by dreamscached** [comment]
Too much trouble to put users through, imo.

**Comment by olivertzeng** [comment]
> Too much trouble to put users through, imo.

true, but still having a way to use this under wayland is still worth it ig. Plus, the user can choose not to

**Comment by olivertzeng** [comment]
> Too much trouble to put users through, imo.

~~not to mention that most linux users are power users~~

serious note that the user could add the flag into the .application file to launch it automatically on every launch

**Comment by dreamscached** [comment]
You'll be surprised how many users come to MAS Discord for tech support with Linux.

**Comment by olivertzeng** [comment]
> You'll be surprised how many users come to MAS Discord for tech support with Linux.

whoops

lmao ||me being arch user moment||
but my perspective is that supporting partially is still better than none ig

**Comment by dreamscached** [comment]
Anyway, the final decision is on the dev team.

**Comment by olivertzeng** [comment]
> Anyway, the final decision is on the dev team.

nice! can't wait!

if it got declined, can i add this feature myself and then send a pr?

**Comment by olivertzeng** [comment]
btw thanks for you and the dev team for the contributions! appreciate it

**Comment by olivertzeng** [comment]
> > https://unix.stackexchange.com/questions/629440/how-to-get-the-url-from-current-tab-of-firefox-without-need-to-install-addition
> 
> > 
> 
> > > I wouldn't use xdotool for that - it is too fragile and error-prone. Use a proper browser automation tool instead such as [Marionette](https://firefox-source-docs.mozilla.org/testing/marionette/Intro.html). Install marionatte_driver python module:
> 
> > > `pip3 install --user marionette_driver`
> 
> > > Start a new instance firefox with --marionette option:
> 
> > > `firefox --marionette`
> 
> > > Use the following get_url.py:
> 
> > > ```python
> 
> > > #!/usr/bin/env python3
> 
> > > 
> 
> > > import marionette_driver
> 
> > > 
> 
> > > m = marionette_driver.marionette.Marionette()
> 
> > > 
> 
> > > m.start_session()
> 
> > > print(m.get_url())
> 
> > > m.delete_session()
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > >     
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >     
> 
> > >   
> 
> > > Example:
> 
> > > ```shell
> 
> > > $ ./get_url.py
> 
> > > https://unix.stackexchange.com/questions/629440/how-to-get-the-url-from-current-tab-of-firefox-by-help-of-wmctrl
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > >     
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >     
> 
> > >   
> 
> > > You can save output of ./get_url.py to a variable using command substitution:
> 
> > > ```shell
> 
> > > $ url="$(./get_url.py )"
> 
> > > $ echo $url
> 
> > > https://unix.stackexchange.com/questions/629440/how-to-get-the-url-from-current-tab-of-firefox-by-help-of-wmctrl/629447?noredirect=1#comment1177925_629447
> 
> > > ```
> 
> > 
> 
> > Kinda cool solution, but apprently only works on gecko browsers
> 
> 
> 
> Browser automation tools don't make sense as a solution to this problem, instead you force users to use a specific instance of the browser that's connected to by the game itself. It *feels* hacky from a user perspective and ultimately can just not play well with user workflows.
> 
> 
> 
> Making the user go through hoops to enable a feature is in most likelihoods a way to make them either get frustrated and give up, or just not want to put in the effort at all. 
> 
> 
> 
> In this case, a partial support solution is worse as it just feels half baked and then supports more niche users, and I just don't think that's worth the effort. 
> 
> 
> 
> We would still reject a PR for these reasons. However, if you wished to make it a submod, you're more than welcome to!

I totally understand that

I'll try to make a submod in the near future for it or maybe find a more robust solution to it IG. Half baked is worse than not having it, ~~just like in ace attorney~~

**Comment by Booplicate** [comment]
I will also note that window reacts are supposed to work for any window, not just browsers.

**Comment by olivertzeng** [comment]
> I will also note that window reacts are supposed to work for any window, not just browsers.

Hmm... alright.

**Comment by olivertzeng** [comment]
Is the Alerts(notifs) feature bundled with the window reaction feature? because just played Just Natsuki today on wayland and noticed that alerts works.

It's prob `notify-send` stuff

**Comment by dreamscached** [comment]
Wayland has nothing to do with notifications.

**Comment by olivertzeng** [comment]
> Wayland has nothing to do with notifications.

Yeah, just as I suspected. But why isn't it supported on Wayland though(I don't see the Alerts section)

**Comment by dreamscached** [comment]
Because when MAS does not detect X11 notifications are disabled.

**Comment by olivertzeng** [comment]
> Because when MAS does not detect X11 notifications are disabled.

wdym

**Comment by olivertzeng** [comment]
is it supposed to mean that mas could not have known if it has notification permission on wayland?

**Comment by dreamscached** [comment]
https://github.com/Monika-After-Story/MonikaModDev/blob/06baf319a34c2ef585bc7c0a1e969a7eaa894b35/Monika%20After%20Story/game/zz_windowutils.rpy#L50-L120

You even mentioned this part of code in your original comment.

I assume if X11 Is not detected MAS disables any notifications, including alerts for topics, not windows.

**Comment by olivertzeng** [comment]
> [`06baf31`/Monika%20After%20Story/game/zz_windowutils.rpy#L50-120](https://github.com/Monika-After-Story/MonikaModDev/blob/06baf319a34c2ef585bc7c0a1e969a7eaa894b35/Monika%20After%20Story/game/zz_windowutils.rpy#L50-120)
> 
> You even mentioned this part of code in your original comment.
> 
> I assume if X11 Is not detected MAS disables any notifications, including alerts for topics, not windows.

Oh wait i almost forgot it lol
why does it do that though, even though that wayland has nothing to do with notifs

**Comment by olivertzeng** [comment]
@dreamscached https://github.com/2e3s/awatcher

I've recently found about ActivityWathcer, a tool that relies on different components `aw-*` like this one called awatcher to record the screentime for users
https://itsfoss.com/activitywatch/
i dont know if this will work for this or not

this watcher has support for wayland

**Comment by olivertzeng** [comment]
![圖片](https://github.com/user-attachments/assets/fb672eac-c041-4e70-a7ad-3c144fd76974)
![圖片](https://github.com/user-attachments/assets/57d5a3f4-a770-450d-b879-176e76b8f23d)

**Comment by olivertzeng** [comment]
it not only works for window titles but also urls as well
and it can even detect if the user is on afk or not

**Comment by dreamscached** [comment]
Looks like a dedicated daemon to run in the background, like it's been said we can't ask users to set up a complex thing like this, we need something that can be included in MAS and work out of the box.

Again, would make a great submod, but imo we can't go that way with vanilla MAS. Asking users for xdotool or notify-send is already a lot.

**Comment by olivertzeng** [comment]
> Looks like a dedicated daemon to run in the background, like it's been said we can't ask users to set up a complex thing like this, we need something that can be included in MAS and work out of the box.
> 
> Again, would make a great submod, but imo we can't go that way with vanilla MAS. Asking users for xdotool or notify-send is already a lot.

It's actually not that hard to implement that way


This just requires installing a Linux package and it can be done with shell scripts

You could even make it do that by including inside the MAS installer

**Comment by olivertzeng** [comment]
> Looks like a dedicated daemon to run in the background, like it's been said we can't ask users to set up a complex thing like this, we need something that can be included in MAS and work out of the box.
> 
> Again, would make a great submod, but imo we can't go that way with vanilla MAS. Asking users for xdotool or notify-send is already a lot.

Or if it's not a supported package manager install the components for them instead

Just a few mv and wgets
Or better if it's included in the /lib directory ig

Just my thoughts though cus awatcher binary isn't big

**Comment by dreamscached** [comment]
It's a daemon that still needs to be taken care of, to launch it, to communicate with it, to dispose of it when MAS crashes (and it crashes a lot) and if MAS crashes there's no way we can reliably take it down.

> Better if it's included in the /lib

IMO better if we'd just use Wayland APIs directly.

**Comment by olivertzeng** [comment]
> It's a daemon that still needs to be taken care of, to launch it, to communicate with it, to dispose of it when MAS crashes (and it crashes a lot) and if MAS crashes there's no way we can reliably take it down.
> 
> 
> 
> > Better if it's included in the /lib
> 
> 
> 
> IMO better if we'd just use Wayland APIs directly.

Yeah that makes sense. Or maybe make the script look something like this

if (awatcher is running || on x11)
    Start the window reactions stuff
And show the window reactions settings if it's installed wether if it's running or not

Like it's a bonus to those who already have installed awatcher

So the way I think to implement this without getting too complicated is to just not include the awatcher binaries and stuff and don't start awatcher explicitly inside the mod

But instead we wait until awatcher is already running alongside MAS

And for some reason if awatcher crashed or just somehow stopped running you don't have close the whole game because you didn't start the watcher from MAS, instead we continue the game and ignore the window reactions stuff

Hope that you understand what I meant

**Comment by olivertzeng** [comment]
As a sidenote implementing this would just need awatcher to be installed, aw-qt is optional because you can just fetch the information of afk and non-afk window names and title names with `aw-awatcher -vvv`

And I believe that this information is recorded inside a log file


So at the end what the mod would try to do if awatcher is running is that it fetches the file and interprets the information on it

**Comment by dreamscached** [comment]
Quoting @multimokia on your previous idea
> It feels hacky from a user perspective and ultimately can just not play well with user workflows.

That's pretty much what *I still myself feel about awatcher, but maybe devs have another opinio.

**Comment by olivertzeng** [comment]
> Quoting @multimokia on your previous idea
> 
> > It feels hacky from a user perspective and ultimately can just not play well with user workflows.
> 
> 
> 
> That's pretty much what *I still* myself feel about awatcher, but maybe devs have another opinion.

Oh well

So I probably should keep this issue open right?

Btw thanks all of you guys trying to reply these issues! I appreciate the work you all have done

**Comment by dreamscached** [comment]
There's still a need in window detection way so unless it's resolved I think it can stay, feel free to suggest other approaches if you found any.

**Comment by olivertzeng** [comment]
> There's still a need in window detection way so unless it's resolved I think it can stay, feel free to suggest other approaches if you found any.

Just asking

What would be the ideal tool to do window detection?

Like if it exists what will it be the properties? That way I can manage to find these tool in the right direction I guess.

Is it like a Python module? Or a daemon with logs? Something like xdotool but for Wayland?

**Comment by dreamscached** [comment]
To me an ideal approach is like Xlib &mdash; a library we can simply include and use in MAS code directly.

**Comment by Booplicate** [comment]
> https://github.com/Monika-After-Story/MonikaModDev/blob/06baf319a34c2ef585bc7c0a1e969a7eaa894b35/Monika%20After%20Story/game/zz_windowutils.rpy#L50-120
> 
> You even mentioned this part of code in your original comment.
> 
> I assume if X11 Is not detected MAS disables any notifications, including alerts for topics, not windows.

Actually, why do we disable alerts on wayland? @multimokia

**Comment by multimokia** [comment]
I'm actually not sure why we do. Really alerts should still run the same way I'm thinking
